tests/components/http/test_forwarded.py

```diff
@@ -0,0 +1,487 @@
+
+
+async def test_x_forwarded_proto_with_trusted_proxy_multiple_for(aiohttp_client):
+    """Test that we get the proto with 1 element in the proto, multiple in the for."""
+
+    async def handler(request):
+        url = mock_api_client.make_url("/")
+        assert request.host == f"{url.host}:{url.port}"
+        assert request.scheme == "https"
+        assert request.secure
+        assert request.remote == "255.255.255.255"
+
+        return web.Response()
+
+    app = web.Application()
+    app.router.add_get("/", handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.0/24")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/",
+        headers={
+            X_FORWARDED_FOR: "255.255.255.255, 127.0.0.1, 127.0.0.2",
+            X_FORWARDED_PROTO: "https",
+        },
+    )
+
+    assert resp.status == 200
+
+
+async def test_x_forwarded_proto_not_processed_without_for(aiohttp_client):
+    """Test that proto header isn't processed without a for header."""
+
+    async def handler(request):
+        url = mock_api_client.make_url("/")
+        assert request.host == f"{url.host}:{url.port}"
+        assert request.scheme == "http"
+        assert not request.secure
+        assert request.remote == "127.0.0.1"
+
+        return web.Response()
+
+    app = web.Application()
+    app.router.add_get("/", handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get("/", headers={X_FORWARDED_PROTO: "https"})
+
+    assert resp.status == 200
+
+
+async def test_x_forwarded_proto_with_multiple_headers(aiohttp_client, caplog):
+    """Test that we get a HTTP 400 bad request with multiple headers."""
+    app = web.Application()
+    app.router.add_get("/", mock_handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/",
+        headers=[
+            (X_FORWARDED_FOR, "222.222.222.222"),
+            (X_FORWARDED_PROTO, "https"),
+            (X_FORWARDED_PROTO, "http"),
+        ],
+    )
+
+    assert resp.status == 400
+    assert "Too many headers for X-Forward-Proto" in caplog.text
+
+
+@pytest.mark.parametrize(
+    "x_forwarded_proto", ["", ",", "https, , https", "https, https, "],
+)
+async def test_x_forwarded_proto_empty_element(
+    x_forwarded_proto, aiohttp_client, caplog
+):
+    """Test that we get a HTTP 400 bad request with empty proto."""
+    app = web.Application()
+    app.router.add_get("/", mock_handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/", headers={X_FORWARDED_FOR: "1.1.1.1", X_FORWARDED_PROTO: x_forwarded_proto},
+    )
+
+    assert resp.status == 400
+    assert "Empty item received in X-Forward-Proto header" in caplog.text
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)